src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,259 @@
+<?php
+
+class WP_HTML_Processor extends WP_HTML_Tag_Processor {
+	public $fully_supported_input = null;
+	public $open_elements = array();
+
+	public function ensure_support() {
+		if ( null !== $this->fully_supported_input ) {
+			return $this->fully_supported_input;
+		}
+
+		$stack = array();
+
+		$p = new WP_HTML_Tag_Processor( $this->html );
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			$tag_name = $p->get_tag();
+
+			if ( ! $p->is_tag_closer() ) {
+				$element = WP_HTML_Spec::element_info( $tag_name );
+
+				$self_closes = $element::is_void || ( ! $element::is_html && $p->has_self_closing_flag() );
+				if ( ! $self_closes ) {
+					$stack[] = $tag_name;
+				}
+			} else {
+				if ( end( $stack ) === $tag_name ) {
+					array_pop( $stack );
+					continue;
+				}
+
+				$this->fully_supported_input = false;
+				return false;
+			}
+		}
+
+		$this->fully_supported_input = 0 === count( $stack );
+
+		return $this->fully_supported_input;
+	}
+
+	public function next_tag( $query = null ) {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		if ( 0 < count( $this->open_elements ) ) {
+			$element = WP_HTML_Spec::element_info( end( $this->open_elements ) );
+			// @TODO: Handle self-closing HTML foreign elements: must convey self-closing flag on stack.
+			if ( $element::is_void ) {
+				array_pop( $this->open_elements );
+			}
+		}
+
+		if ( false === parent::next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			return false;
+		}
+
+		$tag_name = $this->get_tag();
+		$element = WP_HTML_Spec::element_info( $tag_name );
+
+		$self_closes = $element::is_void || ( ! $element::is_html && $this->has_self_closing_flag() );
+		if ( $self_closes ) {
+			$this->open_elements[] = $tag_name;
```

Comment:
I don't like this part, it's not really an open element. Even in the parsing spec these are only added to the stack for brevity of the document – they're always immediately popped in the same step and before ingesting the next token

Reply:
I think it's essential: see above. may not be able to articulate it well yet, but I think one case we might examine is finding the `next_sibling()` of an `<img>`. if we don't push void elements onto the stack we'll be looking for a tag that's a parent of the `img` instead of a sibling.

I think this will matter also for CSS selection. we will need to know the full "tree path" (a term I'm making up here without intending to convey specific information about it) in order to match our current position against the selector.

Reply:
Btw it's not what the spec says, is it? Depending on the insertion mode and element name, we may or may not have to push it to the stack of open elements. There may be other operations to do before pushing, too, and the order matters.